src/wrapper.rs

```diff
@@ -374,3 +389,120 @@ impl Dxc {
+        self.library = DxcLibrary::new(ComPtr::new());
+    }
+}
+
+pub type DxcValidatorVersion = (u32, u32);
+
+// https://www.wihlidal.com/blog/pipeline/2018-09-16-dxil-signing-post-compile/
+impl DxcValidator {
+    fn new(inner: ComPtr<IDxcValidator>, dxc: Dxc, library: DxcLibrary) -> Self {
+        Self {
+            inner,
+            dxc,
+            library,
+        }
+    }
+
+    pub fn version(&self) -> Result<DxcValidatorVersion, HRESULT> {
+        // TODO:
+        /*iid!(pub IID_IDxcVersionInfo = 0xb04f5b50, 0x2059, 0x4f12, 0xa8, 0xff, 0xa1, 0xe0, 0xcd, 0xe1, 0xcc, 0x7e);
+        com_interface! {
+            interface IDxcVersionInfo: IUnknown{
+                iid: IID_IDxcVersionInfo,
+                vtable: IDxcVersionInfoVtbl,
+
+                fn get_version(major: *mut u32, minor: *mut u32) -> HRESULT;
+                fn get_flags(flags: *mut u32) -> HRESULT;
+            }
+        }*/
+        Ok((1, 3))
+    }
+
+    pub fn validate_slice(&mut self, data: &[u8]) -> Result<(Vec<u8>, String), HRESULT> {
+        let blob_encoding = self.library.create_blob_with_encoding(data)?;
+        let blob: DxcBlob = blob_encoding.into();
+        let (blob, errors) = self.validate(blob)?;
+        if errors.is_empty() {
+            Ok((blob.to_vec(), String::new()))
+        } else {
+            Ok((Vec::new(), errors))
+        }
+    }
+
+    pub fn validate(&mut self, blob: DxcBlob) -> Result<(DxcBlob, String), HRESULT> {
+        let mut result: ComPtr<IDxcOperationResult> = ComPtr::new();
+        let result_hr = unsafe {
+            self.inner.validate(
+                blob.inner.as_ptr(),
+                DXC_VALIDATOR_FLAGS_IN_PLACE_EDIT,
+                result.as_mut_ptr(),
+            )
+        };
+        if result_hr != 0 {
+            //Failed to validate dxil container
+            return Err(result_hr);
+        }
+
+        let mut validate_status = 0u32;
+        let result_hr = unsafe { result.get_status(&mut validate_status) };
+        if result_hr != 0 {
+            // Failed to get dxil validate status
+            return Err(result_hr);
+        }
+
+        let (blob, errors) = if validate_status == 0 {
+            (blob, String::new())
+        } else {
+            // The dxil container failed validation
+            let result = DxcOperationResult::new(result);
+            let print_blob = result.get_error_buffer()?;
+            (blob, self.library.get_blob_as_string(&print_blob))
+        };
+
+        Ok((blob, errors))
+    }
+}
+
+#[derive(Debug)]
+pub struct Dxil {
+    dxil_lib: Library,
+}
+
+impl Dxil {
+    pub fn new() -> Self {
+        let dxil_lib = Library::new("dxil.dll").expect("Failed to load dxil.dll");
+        Self { dxil_lib }
+    }
+
+    fn get_dxc_create_instance(&self) -> Symbol<DxcCreateInstanceProc> {
+        unsafe { self.dxil_lib.get(b"DxcCreateInstance\0").unwrap() }
+    }
+
+    pub fn create_validator(&self) -> Result<DxcValidator, HRESULT> {
```

Comment:
Might be nicer to borrow a `library` here instead so we don't end up trying to load the dxcompiler.dll multiple times in case a user already has one laying around.

Not quite sure how I feel about it needing a `Dxc` instance to create a blob, but it seems that might just be the way things are since `dxapi.h` doesn't really expose a different way.